File: mechaverse-simulation/src/test/core/simulation/ant/AntSimulationUtilTest.cpp
#include <gtest/gtest.h>

#include <core/simulation/ant/AntSimulationUtil.h>
#include <core/simulation/ant/ant-simulation.pb.h>

using namespace mechaverse;

TEST(AntSimulationUtil, directionCW) {
  EXPECT_EQ(SOUTH_EAST, AntSimulationUtil::directionCW(EAST));
  EXPECT_EQ(SOUTH, AntSimulationUtil::directionCW(SOUTH_EAST));
  EXPECT_EQ(SOUTH_WEST, AntSimulationUtil::directionCW(SOUTH));
  EXPECT_EQ(WEST, AntSimulationUtil::directionCW(SOUTH_WEST));
  EXPECT_EQ(NORTH_WEST, AntSimulationUtil::directionCW(WEST));
  EXPECT_EQ(NORTH, AntSimulationUtil::directionCW(NORTH_WEST));
  EXPECT_EQ(NORTH_EAST, AntSimulationUtil::directionCW(NORTH));
  EXPECT_EQ(EAST, AntSimulationUtil::directionCW(NORTH_EAST));
}

TEST(AntSimulationUtil, directionCCW) {
  EXPECT_EQ(NORTH_EAST, AntSimulationUtil::directionCCW(EAST));
  EXPECT_EQ(NORTH, AntSimulationUtil::directionCCW(NORTH_EAST));
  EXPECT_EQ(NORTH_WEST, AntSimulationUtil::directionCCW(NORTH));
  EXPECT_EQ(WEST, AntSimulationUtil::directionCCW(NORTH_WEST));
  EXPECT_EQ(SOUTH_WEST, AntSimulationUtil::directionCCW(WEST));
  EXPECT_EQ(SOUTH, AntSimulationUtil::directionCCW(SOUTH_WEST));
  EXPECT_EQ(SOUTH_EAST, AntSimulationUtil::directionCCW(SOUTH));
  EXPECT_EQ(EAST, AntSimulationUtil::directionCCW(SOUTH_EAST));
}

TEST(AntSimulationUtil, canMoveAntToCell_emptyCell) {
  Cell cell = Cell(1, 1);
  EXPECT_TRUE(AntSimulationUtil::canMoveAntToCell(cell));
}

TEST(AntSimulationUtil, canMoveAntToCell_rockCell) {
  Cell cell = Cell(1, 1);
  Rock rock = Rock(Rock::default_instance());
  cell.setRock(&rock);
  EXPECT_FALSE(AntSimulationUtil::canMoveAntToCell(cell));
}

File: simulation/ant/mechaverse-simulation-ant-core/src/test/java/org/mechaverse/simulation/ant/core/AntSimulationUtilTest.java
package org.mechaverse.simulation.ant.core;

import static org.junit.Assert.assertEquals;

import org.junit.Test;
import org.mechaverse.simulation.ant.core.model.Ant;
import org.mechaverse.simulation.common.model.Direction;
import org.mechaverse.simulation.common.util.SimulationUtil;

/**
 * Unit tests for {@link AntSimulationUtil}.
 */
public class AntSimulationUtilTest {

  @Test
  public void testTurnCW() {
    Ant ant = new Ant();
    for(Direction direction : Direction.values()) {
      ant.setDirection(direction);
      SimulationUtil.turnCW(ant);
      assertEquals(SimulationUtil.directionCW(direction), ant.getDirection());
    }
  }

  @Test
  public void testTurnCCW() {
    Ant ant = new Ant();
    for(Direction direction : Direction.values()) {
      ant.setDirection(direction);
      SimulationUtil.turnCCW(ant);
      assertEquals(SimulationUtil.directionCCW(direction), ant.getDirection());
    }
  }

  @Test
  public void directionCW() {
    assertEquals(Direction.SOUTH_EAST, SimulationUtil.directionCW(Direction.EAST));
    assertEquals(Direction.SOUTH, SimulationUtil.directionCW(Direction.SOUTH_EAST));
    assertEquals(Direction.SOUTH_WEST, SimulationUtil.directionCW(Direction.SOUTH));
    assertEquals(Direction.WEST, SimulationUtil.directionCW(Direction.SOUTH_WEST));
    assertEquals(Direction.NORTH_WEST, SimulationUtil.directionCW(Direction.WEST));
    assertEquals(Direction.NORTH, SimulationUtil.directionCW(Direction.NORTH_WEST));
    assertEquals(Direction.NORTH_EAST, SimulationUtil.directionCW(Direction.NORTH));
    assertEquals(Direction.EAST, SimulationUtil.directionCW(Direction.NORTH_EAST));
  }

  @Test
  public void testDirectionCCW() {
    assertEquals(Direction.NORTH_EAST, SimulationUtil.directionCCW(Direction.EAST));
    assertEquals(Direction.NORTH, SimulationUtil.directionCCW(Direction.NORTH_EAST));
    assertEquals(Direction.NORTH_WEST, SimulationUtil.directionCCW(Direction.NORTH));
    assertEquals(Direction.WEST, SimulationUtil.directionCCW(Direction.NORTH_WEST));
    assertEquals(Direction.SOUTH_WEST, SimulationUtil.directionCCW(Direction.WEST));
    assertEquals(Direction.SOUTH, SimulationUtil.directionCCW(Direction.SOUTH_WEST));
    assertEquals(Direction.SOUTH_EAST, SimulationUtil.directionCCW(Direction.SOUTH));
    assertEquals(Direction.EAST, SimulationUtil.directionCCW(Direction.SOUTH_EAST));
  }
}
